modules/imgcodecs/test/test_tiff.cpp

```diff
@@ -219,6 +219,46 @@ TEST(Imgcodecs_Tiff, readWrite_32FC3_RAW)
     EXPECT_EQ(0, remove(filenameOutput.c_str()));
 }
 
+TEST(Imgcodecs_Tiff, readWrite_predictor)
+{
+    /* see issue #21871
+     */
+    const uchar sample_data[160] = {
+        0xff, 0xff, 0xff, 0xff, 0x88, 0x88, 0xff, 0xff, 0x88, 0x88, 0xff, 0xff, 0xff, 0xff, 0xff, 0x88,
+        0xff, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00,
+        0xff, 0x00, 0x00, 0x44, 0xff, 0xff, 0x88, 0xff, 0x33, 0x00, 0x66, 0xff, 0xff, 0x88, 0x00, 0x44,
+        0x88, 0x00, 0x44, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x44, 0xff, 0xff, 0x11, 0x00, 0xff,
+        0x11, 0x00, 0x88, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0xff, 0xff, 0x00, 0x00, 0xff,
+        0x11, 0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x33, 0x00, 0x88, 0xff, 0x00, 0x66, 0xff,
+        0x11, 0x00, 0x66, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x44, 0x33, 0x00, 0xff, 0xff,
+        0x88, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff,
+        0xff, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0x33, 0x00, 0x00, 0x66, 0xff, 0xff,
+        0xff, 0xff, 0x88, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0xff, 0xff, 0xff
+    };
+
+    cv::Mat mat(10, 16, CV_8UC1, (void*)sample_data);
+    int methods[] = {
+        COMPRESSION_NONE,     COMPRESSION_LZW,      COMPRESSION_JPEG,
+        COMPRESSION_PACKBITS, COMPRESSION_DEFLATE,  COMPRESSION_ADOBE_DEFLATE
+    };
+    for (size_t i = 0; i < sizeof(methods) / sizeof(int); i++)
+    {
+        string out = cv::tempfile(".tif");
+
+        std::vector<int> params;
+        params.push_back(TIFFTAG_COMPRESSION);
+        params.push_back(methods[i]);
+        params.push_back(TIFFTAG_PREDICTOR);
+        params.push_back(PREDICTOR_HORIZONTAL);
+
+        EXPECT_NO_THROW(cv::imwrite(out, mat, params));
+
+        const Mat img = cv::imread(out, IMREAD_UNCHANGED);
+        ASSERT_FALSE(img.empty());
```

Comment:
It makes sense to compare image data after decoding like `ASSERT_EQ(0, cv::countNonZero(cv::absdiff(mat, img)));`.

Reply:
Exactly. It makes sense as long as it is a lossless compression. I remove `COMPRESSION_JPEG` from the testing code in that it is a lossy one.